text/000-rustdoc-bundle-local-resources.md

```diff
@@ -0,0 +1,92 @@
+Rustdoc: Bundle local images
+
+- Feature Name: NONE
+- Start Date: 2023-02-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#32104](https://github.com/rust-lang/rust/issues/32104)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to allow the bundling of local images in rustdoc HTML output. A draft implementation is available as [#107640](https://github.com/rust-lang/rust/pull/107640).
+
+# Motivation
+[motivation]: #motivation
+
+Currently, rustdoc does not allow for the inclusion of local images in the generated output. This limits users from customizing their rustdoc output by adding images to their documentation as they currently need to host them on a web server so they can be available from anywhere, but therefore requiring an access to the internet. Bundling local images would enable users to customize their rustdoc output and make it more visually appealing.
+
+This would make the documentation more engaging and easier to understand while lowering the amount of effort required to achieve a better result.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC proposes to allow rustdoc to include local images in the generated documentation by copying them into the output directory.
+
+This would be done by allowing users to specify the path of a local resource file in doc comments. The resource file would be stored in the `doc.files` folder. The `doc.files` folder will be at the "top level" of the rustdoc output level (at the same level as the `static.files` or the `src` folders).
+
+The only local resources considered will be the ones in the markdown image syntax: `![resource title](path)`, where `<path>` is the path of the resource file relative to the source file.
+
+The path could be either a relative path (`../images/my_image.png`) or an absolute path (like `/home/user/project/images/my_image.png` or `C:/Users/user/project/images/my_image.png`) so that paths can be constructed using `OUT_DIR`:
+
+```rust
+/// Using a local image
+#[doc=concat!("![with absolute path](", env!("OUT_DIR"), "/local/image.png)")]
+///
+/// Using a local image ![with relative path](../local/image.png)
```

Comment:
Can't you already do this today by using `#[doc = include_str!("../local/image.png")]`? Why does it need a new feature?

The only thing I've seen mentioned so far is that it would allow outlining the contents of the attribute into its own file, but that just seems like a performance enhancement? Is the performance actually causing issues in practice?

Reply:
(Though this current RFC doesn't support favicons and crate logo; I think it would be worth adding those at the same time since they are the primary usecase and the semantics are the same, it's just a little extra syntax question of how we detect whether it's a URL or file path).

Reply:
The logo and favicon are definitely targets of this RFC. I'm just not sure if we want to put it directly into this RFC or if we should make a follow-up PR. I'll add it into `Unresolved questions`.